app/models/concerns/message_subscription.rb

```diff
@@ -1,87 +1,11 @@
-#
+# frozen_string_literal: true
+
 # Logic for which users to subscribe to each new message
 module MessageSubscription
   extend ActiveSupport::Concern
 
   # Who gets subscribed to a new message (i.e. who can see each message)
   def subscribe_users_to_message
-    obj = self.related_object
-    klass = obj.class.name.underscore
-    user_ids = []
-
-    # Add the following users
-    #   - Donor / Charity user
-    #   - Message sender
-    #   - Anyone who has previously replied to offer/order
-    #   - Admin users processing the offer/order
-    user_ids << obj.try(:created_by_id)
-    user_ids << self.sender_id
-    user_ids += public_subscribers_to(klass, obj.id)
-    user_ids += private_subscribers_to(klass, obj.id)
-    admin_user_fields.each{|field| user_ids << obj.try(field)}
-
-    # Remove the following users
-    #   - SystemUser and StockitUser
-    #   - donor/charity user if the message is private (supervisor channel) or offer/order is cancelled
-    user_ids = user_ids.flatten.uniq
-    user_ids -= [User.system_user.try(:id), User.stockit_user.try(:id)]
-    user_ids -= [obj.try(:created_by_id)] if self.is_private or obj.try('cancelled?')
-
-    # Cases where we subscribe every staff member
-    #  - For private messages, subscribe all supervisors ONLY for the first message
-    #  - If donor sends a message but no one else is listening, subscribe all reviewers.
-    subscribe_all_staff =
-      if is_private
-        is_first_message_for(klass, obj.id)
-      else
-        obj&.created_by_id.present? && (user_ids == [self.sender_id])
-      end
-
-    user_ids += User.staff.pluck(:id) if subscribe_all_staff
-
-    user_ids.flatten.compact.uniq.each do |user_id|
-      state = (user_id == self.sender_id) ? "read" : "unread" # mark as read for sender
-      add_subscription(state, user_id)
-    end
-  end
-
-  private
-
-  # A public subscriber is defined as :
-  #   > Anyone who has a subscription to that record
-  def public_subscribers_to(klass, id)
-    Subscription
-      .joins(:message)
-      .where("#{klass}_id": id, messages: { is_private: false })
-      .pluck(:user_id)
+    Messages::Operations.new(message: self).subscribe_users_to_message
```

Comment:
I like scoping the operations into its own namespace.

But there's no need to instantiate an object if it gets thrown out immediately after. You can use class methods here

`Messages::Operations.subscribe_users_to_message(message: self)`

Reply:
@patrixr , cannot use class method here, as `subscribe_users_to_message` is utilising other methods that are instance methods.
and also, the initilization is important for `subscribe_users_to_message`

Reply:
Those other instance methods would also become class methods

Reply:
i have refactored this even a bit more. i'll be removing this completely

Reply:
Cool, let me know when it's updated.

As a general rule, we don't want to instanciate anything that will only be used to call a method once.

```ruby
# Instanciated version

class MyService
  def do_something; end
end

MyService.new.do_something # Not a good way to expose the entry point

# Static version

class MyService
  def self.do_something; end
end

MyService.do_something # Better way to expose the entry point

# Hybrid version

class MyService
  def self.do_something
    MyService.new.do_something
  end

  def do_something; end
end

MyService.do_something # Same as above, except it's just a syntaxic sugar and you can still instanciate it if more operations are required

```

Reply:
this has been done : )